src/visualization/src/Texture.cpp

```diff
@@ -0,0 +1,116 @@
+{
+    return textureEnvironment;
+}
+
+iDynTree::ColorViz iDynTree::Texture::getPixelColor(unsigned int width, unsigned int height) const
+{
+    ColorViz pixelOut;
+    irr::video::SColor pixelIrrlicht = irr::video::SColor(0, 0, 0, 0);
+
+    if (!irrTexture)
+    {
+        reportError("Texture","getTexturePixelColor","Cannot get pixel color. The video texture has not been properly initialized.");
+        return pixelOut;
+    }
+
+    auto textureDim = irrTexture->getSize();
+
+    if ( width >= textureDim.Width || height >= textureDim.Height)
+    {
+        std::stringstream ss;
+        ss << "The requested pixel is out of bounds. Requested (" << width << ", " << height
+           << "). Picture dimensions: (" << textureDim.Width  << ", " << textureDim.Height << ").";
+        reportError("Visualizer", "getTexturePixelColor", ss.str().c_str());
+        return pixelOut;
+    }
+
+    auto pitch = irrTexture->getPitch();
+    auto format = irrTexture->getColorFormat();
+    auto bytes = irr::video::IImage::getBitsPerPixelFromFormat(format) / 8;
+
+    unsigned char* buffer = (unsigned char*) irrTexture->lock(irr::video::E_TEXTURE_LOCK_MODE::ETLM_READ_ONLY);
+    if (buffer)
+    {
+        pixelIrrlicht = irr::video::SColor(*(unsigned int*)(buffer + (height * pitch) + (width * bytes)));
+        irrTexture->unlock();
+    }
+
+    pixelOut.r = pixelIrrlicht.getRed();
+    pixelOut.g = pixelIrrlicht.getGreen();
+    pixelOut.b = pixelIrrlicht.getBlue();
+    pixelOut.a = pixelIrrlicht.getAlpha();
+
+    return pixelOut;
+}
+
+bool iDynTree::Texture::getPixels(std::vector<iDynTree::PixelViz> &pixels) const
+{
+    irr::video::SColor pixelIrrlicht = irr::video::SColor(0, 0, 0, 0);
+
+    if (!irrTexture)
+    {
+        reportError("Texture","getTexturePixels","Cannot get pixel color. The video texture has not been properly initialized.");
+        return false;
+    }
+
+    auto textureDim = irrTexture->getSize();
+
+    pixels.resize(textureDim.Width * textureDim.Height);
+
+    auto pitch = irrTexture->getPitch();
+    auto format = irrTexture->getColorFormat();
+    auto bytes = irr::video::IImage::getBitsPerPixelFromFormat(format) / 8;
+
+    unsigned char* buffer = (unsigned char*) irrTexture->lock(irr::video::E_TEXTURE_LOCK_MODE::ETLM_READ_ONLY);
+    if (buffer)
+    {
+        size_t i = 0;
+        for (size_t width = 0; width < textureDim.Width; ++width)
+        {
+            for (size_t height = 0; height < textureDim.Height; ++height)
+            {
+                pixelIrrlicht = irr::video::SColor(*(unsigned int*)(buffer + (height * pitch) + (width * bytes)));
+                pixels[i].width = width;
+                pixels[i].height = height;
+                pixels[i].r = pixelIrrlicht.getRed();
+                pixels[i].g = pixelIrrlicht.getGreen();
+                pixels[i].b = pixelIrrlicht.getBlue();
+                pixels[i].a = pixelIrrlicht.getAlpha();
+                ++i;
+            }
+        }
```

Comment:
FYI, this is pretty inefficient. I would have liked to output directly ``buffer``, but it should have been necessary to implement also the lock/unlock mechanism required by ``irrlicht``. In addition, it was not clear how to pass the information on how to interpret that buffer. For RGB images, having the raw buffer may be handy with ``YARP`` images, thus avoiding two double for loops. In fact, ``irrlicht`` and ``YARP`` seem to store RGB images in the same way. It is not the same for RGBA. In fact, ``irrlicht`` uses ARGB instead.
Bottom line is, I did not find a better way to output the raw image. In any case, this has an effect only when the corresponding texture is large, or when getting the pixels of many textures.